cpp/src/sort/sort_impl.cuh

```diff
@@ -0,0 +1,497 @@
+// -*-c++-*-
+
+/*
+ * Copyright (c) 2019, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#pragma once
+
+#include <stdlib.h>
+#include <omp.h>
+
+#include "binning.cuh"
+#include <cub/cub.cuh>
+
+#include <cub/device/device_radix_sort.cuh>
+#include <cub/util_allocator.cuh>
+
+#include <thrust/sort.h>
+#include <thrust/for_each.h>
+#include <thrust/execution_policy.h>
+
+#include "utilities/error_utils.h"
+#include "rmm_utils.h"
+
+namespace cusort {
+
+  namespace detail {
+    //
+    //  Define a device function to count leading zeros, since
+    //  the intrinsic is different for each type.
+    //
+    //  Note, C++ doesn't currently support partial template
+    //  specialization, so this is done with a function object.
+    //
+    template <typename Key_t, int size>
+    struct CountLeadingZeros {
+      __inline__ __device__ int operator()(Key_t k) {
+        return __clz(k);
+      }
+    };
+
+    template <typename Key_t>
+    struct CountLeadingZeros<Key_t, 8> {
+      __inline__ __device__ int operator()(Key_t k) {
+        return __clzll(k);
+      }
+    };
+  }
+  
+  template <typename Key_t,typename Value_t, typename Length_t,
+            int MAX_NUM_GPUS = 16, int BIN_SCALE = 16,
```

Comment:
@seunghwak , the cusort class itselft doesn't make any "hardcoded" assumptions. These are runtime configurable.

Reply:
I got this but I am not sure this applies to sort_test.cu line 32

```
#define MAX_NUM_GPUS 16
```

as well.

```
TEST_F(SortTest, Random10MPerDevice_uint64_t)
{
  cudaStream_t stream{nullptr};
  
  uint64_t           *d_input[MAX_NUM_GPUS];<= MAX_NUM_GPUS hard coded to 16
  uint64_t           *d_input_values[MAX_NUM_GPUS];
  uint64_t           *d_output[MAX_NUM_GPUS];
  uint64_t           *d_output_values[MAX_NUM_GPUS];
  unsigned long long  h_input_offsets[3];
  unsigned long long  h_output_offsets[3];

  const int num_elements = 10000000;
  const int seed = 43;
  int n_gpus = 0;

  CUDA_RT_CALL(cudaGetDeviceCount(&n_gpus));<= n_gpus can be larger than 16 in future systems.

  for (int i = 0 ; i < (n_gpus + 1) ; ++i)
    h_input_offsets[i] = i * num_elements;<=h_input_offsets's size is set to three, so isn't this a bug as well?

  omp_set_num_threads(n_gpus);

  generate_random(d_input, d_input_values, h_input_offsets, n_gpus, seed, stream);

  //
  //  Initialize Peer-to-peer communications
  //
  cusort::initialize_snmg_communication(n_gpus);

  // NOTE:  could vary numBins, binScale, useThrust
  cusort::sort_key_value(d_input, d_input_values, h_input_offsets,
                         d_output, d_output_values, h_output_offsets,
                         n_gpus);

  verify_sorted_order(d_output, d_output_values, h_output_offsets, n_gpus, stream, true);

  for (int i = 0 ; i < n_gpus ; ++i) {
    cudaSetDevice(i);

    EXPECT_EQ(RMM_FREE(d_input[i], stream), RMM_SUCCESS);<=d_input[i] may not be valid if n_gpus > MAX_NUM_GPUS
    EXPECT_EQ(RMM_FREE(d_output[i], stream), RMM_SUCCESS);
  }
}
```

See the comments starting with <= in the above code.

Reply:
See above discussion.